docs/design/transactionhandling.rst

```diff
@@ -0,0 +1,144 @@
+--------------------
+Flow of Transactions
+--------------------
+
+Weave implements the complexity of the ABCI interface
+for you and only exposes a few key points for you to add
+your custom logic. We provide you a `default merklized
+key value store <https://github.com/iov-one/weave/blob/master/store/iavl/adapter.go>`__ to store all the data, which exposes
+a simple interface, similar to LevelDB.
+
+When you create a `new BaseApp
+<https://github.com/iov-one/weave/blob/master/app/base.go#L22-L33>`__, you must provide:
+
+* a merkelized data store (default provided)
+* a txdecoder to parse the incoming transaction bytes
+* a handler that processes ``CheckTx`` and ``DeliverTx`` (like ``http.Handler``)
+* and optionally a ``Ticker`` that is called every ``BeginBlock`` if you have repeated tasks.
+
+The merkelized data store automatically supports ``Querys``
+(with proofs), and the initial handshake to sync with
+tendermint on startup.
+
+Transactions
+============
+
+A transaction must be `Persistent <#Persistence>`__ and
+contain the message we wish to process, as well as an
+envelope. It implements the minimal ``Tx`` interface,
+and can also implement a number of additional
+interfaces to be compatible with the particular middleware
+stack in use in your application. For example, supporting
+the `x/sigs/Decorator <https://github.com/iov-one/weave/blob/master/x/sigs/decorator.go#L53>`__ 
+or the `x/cash/FeeDecorator <https://github.com/iov-one/weave/blob/master/x/cash/staticfee.go#L114>`__
+require a Tx that fulfills interfaces to expose the signer
+or the fee information.
+
+Once the transaction has been processed by the middleware
+stack, we can call ``GetMsg()`` to extract the actual
+message with the action we wish to perform.
+
+Handler
+=======
+
+As mentioned above, we pass every ``Tx`` through a middleware
+stack to perform standard processing and checks on all
+transactions. However, only the ``Tx`` is validated, we need
+to pass the underlying message to the specific code to handle
+this action.
+
+We do so by taking inspiration from standard http Routers.
+Every message object must implement ``Path()`` , which
+returns a string used by the ``Router`` in order
+to find the proper ``Handler``. The ``Handler`` is
+then responsible for processing any message type that
+is registered with it.
+
+.. code-block:: go
+
+    type Handler interface {
+        Check(ctx Context, store KVStore, tx Tx) (CheckResult, error)
+        Deliver(ctx Context, store KVStore, tx Tx) (DeliverResult, error)
```

Comment:
The notation has changed. Results are now pointers.

I think it is good to use `weave.` package prefix. Otherwise one may think that `Context` is `context.Context` and not `weave.Context` (I know it's an alias :smirk:).

Reply:
Or not? Since it's an alias to actual `Context` from stdlib it might actually be more straightforward for devs that are familiar with Go, as it's easier to understand what it does. A matter of preference I suppose.